### PR TITLE
Documentation template for load balancer Virtual Server

### PR DESCRIPTION
**Pull request by Pushpalatha555**
---
layout: ""
page_title: "hpegl_vmaas_load_balancer_virtual_server Resource - vmaas-terraform-resources"
subcategory: "vmaas"
description: |-
    loadbalancer Virtual Server resource facilitates creating, updating
          and deleting NSX-T Network Load Balancer Virtual Servers.
---

# Resource hpegl_vmaas_load_balancer_virtual_server

-> Compatible version >= 5.4.6

loadbalancer Virtual Server resource facilitates creating, updating
		and deleting NSX-T Network Load Balancer Virtual Servers.
`hpegl_vmaas_load_balancer_virtual_server` resource supports NSX-T Load Balancer Virtual Server creation.

## Example usage for creating NSX-T Load Balancer Virtual Server with all possible attributes

```terraform
# (C) Copyright 2022 Hewlett Packard Enterprise Development LP

resource "hpegl_vmaas_load_balancer_virtual_server" "tf_lb_virtual_server" {
  lb_id       = data.hpegl_vmaas_load_balancer.tf_lb.id
  name        = "tf_virtual-server"
  description = "tf_virtual-server created by tf"
  vip_address = "10.11.12.13"
  vip_port    = "8080"
  pool        = data.hpegl_vmaas_load_balancer_pool.tf_pool.id

  type = "http"
  http_application_profile {
    application_profile = data.hpegl_vmaas_load_balancer_profile.tf_http_profile.id
  }

  persistence = "COOKIE"
  cookie_persistence_profile {
    persistence_profile = data.hpegl_vmaas_load_balancer_profile.tf_cookie_profile.id
  }

  ssl_server_cert = data.hpegl_vmaas_load_balancer_virtual_server_ssl_cert.tf_ssl_cert.id
  ssl_server_config {
    ssl_server_profile = data.hpegl_vmaas_load_balancer_profile.tf_ssl_server_profile.id
  }

  ssl_client_cert = data.hpegl_vmaas_load_balancer_virtual_server_ssl_cert.tf_ssl_cert.id
  ssl_client_config {
    ssl_client_profile = data.hpegl_vmaas_load_balancer_profile.tf_ssl_client_profile.id
  }

}
```

### Required

- `lb_id` (Number) Parent lb ID, lb_id can be obtained by using LB datasource/resource.
- `name` (String) Name of Network loadbalancer virtual server name
- `pool` (Number) Pool Id, Get the `id` from hpegl_vmaas_load_balancer_pool datasource to obtain the Pool Id, It is recommended that you attach a pool to the Virtual Server to have a correct LB functionality
- `ssl_client_cert` (Number) ssl_client_cert Id, Get the `id` from hpegl_vmaas_load_balancer_virtual_server_ssl_cert datasource to obtain the ssl_client_cert Id, SSLClientCert is needed only for https based load balancer
- `ssl_server_cert` (Number) ssl_server_cert Id, Get the `id` from hpegl_vmaas_load_balancer_virtual_server_ssl_cert datasource to obtain the ssl_server_cert Id, SSLServerCert is needed only for https based load balancer
- `type` (String) Vip protocol of Network loadbalancer virtual server
- `vip_address` (String) Vip_address of Network loadbalancer virtual server
- `vip_port` (String) Vip_port of network loadbalancer virtual server

### Optional

- `cookie_persistence_profile` (Block List, Max: 1) Cookie profile configuration (see [below for nested schema](#nestedblock--cookie_persistence_profile))
- `description` (String) Description of Network loadbalancer virtual server
- `http_application_profile` (Block List, Max: 1) HTTP Profile configuration (see [below for nested schema](#nestedblock--http_application_profile))
- `persistence` (String) Persistence type for Network loadbalancer virtual server
- `sourceip_persistence_profile` (Block List, Max: 1) HTTP profile configuration (see [below for nested schema](#nestedblock--sourceip_persistence_profile))
- `ssl_client_config` (Block List) virtual server Configuration (see [below for nested schema](#nestedblock--ssl_client_config))
- `ssl_server_config` (Block List) virtual server Configuration (see [below for nested schema](#nestedblock--ssl_server_config))
- `tcp_application_profile` (Block List, Max: 1) TCP Profile configuration (see [below for nested schema](#nestedblock--tcp_application_profile))
- `udp_application_profile` (Block List, Max: 1) UDP profile configuration (see [below for nested schema](#nestedblock--udp_application_profile))

### Read-Only

- `id` (String) The ID of this resource.

<a id="nestedblock--cookie_persistence_profile"></a>
### Nested Schema for `cookie_persistence_profile`

Required:

- `persistence_profile` (Number) COOKIE persistence_profile Id, Get the `id` from hpegl_vmaas_load_balancer_profile datasource to obtain the COOKIE persistence_profile Id

<a id="nestedblock--http_application_profile"></a>
### Nested Schema for `http_application_profile`

Required:

- `application_profile` (Number) HTTP application_profile Id, Get the `id` from hpegl_vmaas_load_balancer_profile datasource to obtain the HTTP application_profile Id

<a id="nestedblock--sourceip_persistence_profile"></a>
### Nested Schema for `sourceip_persistence_profile`

Required:

- `persistence_profile` (Number) SOURCEIP persistence_profile Id, Get the `id` from hpegl_vmaas_load_balancer_profile datasource to obtain the SOURCEIP persistence_profile Id

<a id="nestedblock--ssl_client_config"></a>
### Nested Schema for `ssl_client_config`

Required:

- `ssl_client_profile` (Number) ssl_client_profile Id, Get the `id` from hpegl_vmaas_load_balancer_profile datasource to obtain the ssl_client_profile Id

<a id="nestedblock--ssl_server_config"></a>
### Nested Schema for `ssl_server_config`

Required:

- `ssl_server_profile` (Number) ssl_server_profile Id, Get the `id` from hpegl_vmaas_load_balancer_profile datasource to obtain the ssl_server_profile Id

<a id="nestedblock--tcp_application_profile"></a>
### Nested Schema for `tcp_application_profile`

Required:

- `application_profile` (Number) TCP application_profile Id, Get the `id` from hpegl_vmaas_load_balancer_profile datasource to obtain the TCP application_profile Id

<a id="nestedblock--udp_application_profile"></a>
### Nested Schema for `udp_application_profile`

Required:

- `application_profile` (Number) UDP application_profile Id, Get the `id` from hpegl_vmaas_load_balancer_profile datasource to obtain the UDP application_profile Id

### CODE REVIEW

File: templates/resources/vmaas_load_balancer_virtual_server.md.tmpl [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+---
+
+# Resource hpegl_vmaas_load_balancer_virtual_server
+
+-> Compatible version >= 5.4.6
+
+{{ .Description | trimspace }}
+`hpegl_vmaas_load_balancer_virtual_server` resource supports NSX-T Load Balancer Virtual Server creation.
+
+->  Load Balancer ID Data Source `hpegl_vmaas_load_balancer` is supported for Creating Virtual Server.
```
Comment (manjunath-batakurki) [inline templates/resources/vmaas_load_balancer_virtual_server.md.tmpl:16]: this should ideally come in your field description. U cannot add a note for saying there is a data Source. Please refer existing documents.

https://registry.terraform.io/providers/HPE/hpegl/latest/docs/resources/vmaas_network

Comment (Pushpalatha555) [reply templates/resources/vmaas_load_balancer_virtual_server.md.tmpl]: removed

File: internal/resources/resource_lb_virtual_servers.go [hunk 2]
```diff
@@ -74,14 +70,15 @@ func LoadBalancerVirtualServers() *schema.Resource {
...
 			"cookie_persistence_profile":   schemas.CookiePersProfileSchema(),
 			"sourceip_persistence_profile": schemas.SourceipPersProfileSchema(),
 			"ssl_server_cert": {
-				Type:        schema.TypeInt,
-				Required:    true,
-				Description: "ID of the ssl_server_cert. Use " + DSLBVirtualServerSslCert + "datasource to obtain the id  here",
+				Type:     schema.TypeInt,
+				Required: true,
+				Description: "SSLServerCert is needed only for https based load balancer" +
+					"ID of the ssl_server_cert. Use " + DSLBVirtualServerSslCert + "datasource to obtain the id  here",
```
Comment (manjunath-batakurki) [inline internal/resources/resource_lb_virtual_servers.go:81]: Can you make these statements more use readable ? the grammar is totally messed up here. applies to all the places. I would suggest u to read it before reaching out for review. This comment applies to all the places.

Comment (Pushpalatha555) [reply internal/resources/resource_lb_virtual_servers.go]: updated

**Review by manjunath-batakurki — APPROVED** [review]

**Review by ravikumar29524 — APPROVED** [review]

**Pull request merged**